src/main/java/org/thoughtcrime/securesms/service/FetchForegroundService.java

```diff
@@ -0,0 +1,79 @@
+package org.thoughtcrime.securesms.service;
+
+import android.app.Notification;
+import android.app.Service;
+import android.content.Context;
+import android.content.Intent;
+import android.os.IBinder;
+import android.util.Log;
+
+import androidx.annotation.Nullable;
+import androidx.core.app.NotificationCompat;
+import androidx.core.content.ContextCompat;
+
+import org.thoughtcrime.securesms.ApplicationContext;
+import org.thoughtcrime.securesms.R;
+import org.thoughtcrime.securesms.notifications.FcmReceiveService;
+import org.thoughtcrime.securesms.notifications.NotificationCenter;
+import org.thoughtcrime.securesms.util.Util;
+
+public final class FetchForegroundService extends Service {
+  private static final String TAG = FcmReceiveService.class.getSimpleName();
+  private static final Object SERVICE_LOCK = new Object();
+  private static Intent service;
+
+  public static void start(Context context) {
+    GenericForegroundService.createFgNotificationChannel(context);
+    synchronized (SERVICE_LOCK) {
+      if (service == null) {
+        service = new Intent(context, FetchForegroundService.class);
+        ContextCompat.startForegroundService(context, service);
+      }
+    }
+  }
+
+  public static void stop(Context context) {
+    synchronized (SERVICE_LOCK) {
+      if (service != null) {
+        context.stopService(service);
+        service = null;
+      }
+    }
+  }
+
+  @Override
+  public void onCreate() {
+    Log.i(TAG, "Creating fetch service");
+    super.onCreate();
+
+    Notification notification = new NotificationCompat.Builder(this, NotificationCenter.CH_GENERIC)
+      .setContentTitle(getString(R.string.connectivity_updating))
+      .setSmallIcon(R.drawable.notification_permanent)
+      .build();
+
+    startForeground(NotificationCenter.ID_FETCH, notification);
+
+    // Start explicit fetch only after we marked ourselves as requiring foreground;
+    // this may help we on getting network and time adequately
+    // Fetch is started in background to not block the UI.
+    // We then run not longer than the max. of 20 seconds,
+    // see https://firebase.google.com/docs/cloud-messaging/android/receive .
```

Comment:
I understood https://firebase.google.com/docs/cloud-messaging/android/receive differently:
- FirebaseMessagingService "should handle any message within 20 seconds of receipt", i.e. onMessageReceived should return within 20s
- "The time window for handling a message may be shorter than 20 seconds"
- Therefore, in the code example, they try to need only 10 seconds to be on the safe side:

```
        if (/* Check if data needs to be processed by long running job */ true) {
            // For long-running tasks (10 seconds or more) use WorkManager.
            scheduleJob();
        } else {
            // Handle message within 10 seconds
            handleNow();
        }
```

- As soon as we are in a foreground service, we may need as long as we need to (within some bounds, IIRC on newer Androids you told me it's 6 hours per day or until the user swipes away the notification, anyway, it's way more than 20s)

Conclusion: IIUC, we can set the timeout to more than 20s, and I personally think that we should because on a slow network or so it may be nice to have a minute or even more.

Reply:
> As soon as we are in a foreground service, we may need as long as we need to

indeed, that makes sense to me as well. once there is a foreground service notification, one is usually allowed to run long times - things are visible to the user. only without a foreground service notification one has 20 seconds - and maybe we got hit by that. though, still a bit confusing in the documentation ... 

what would be a reasonable timeout for slow networks then? 60 seconds? the service is not started a second time anyways in case another push arrives and we're not completed.

Reply:
> Yes, I also wrote in the issue that datasync foreground task is allowed to run for 6 hours or so.

IIRC  it is 6 hours for the whole day, so if you use it for several minutes in every push notification you might run out of "quota"

Reply:
i changed the timeout to 300 seconds in the last commit and adapted the comment

Reply:
> The "ideal" solution that only requires a foreground service (with "Updating..." notification) when necessary

yeah, that may or may not work, probably hard to say without quite some testing across different android versions, manufactors etc.

might be, systems require foreground service and grant network access only there. at least, this was the reasoning to move backgroundFetch() inside FetchForegroundService above.

so, i think, the current approach is kind of more bullet proof as it has less state and is pretty straight forward. let's see if the additional notification is really annoying and visible in case there is good connectivity